### PR TITLE
Show carousel of images when multiple images are available

### LINKED ISSUE #13981: [Shopfront - Product details] Show carousel of images when multiple images are available
This issue enables shoppers to see multiple images for a product, as enabled in the admin by #13978.

## Description
**- As a:** shopper
**- On page:** /shop#/shop_panel > product overlay
**- I want to be able to do:** see multiple photos of a product

## Acceptance Criteria & Tests

1. Don't show carousel
Given there is a products available in a shop
And that product doesn't have multiple images
When I click to see the product details
Then I only see one image (current behaviour)

2. Show carousel
Given there is a products available in a shop
And that product has multiple images
When I click to see the product details
Then I see a carousel with multiple images
And the order is based on the order of the variants

3. Show carousel in product preview (admin)
Given I have a product with multiple images
When I click to see its preview > product details
Then I see a carousel with multiple images
And the order is based on the order of the variants

Image below is example only. Depending on what carousel we decide to use, we will tweak based on the available options.

Image

### PR DESCRIPTION
**Pull request by chahmedejaz**
**:warning: Please use clockify #82 - Multiple images when working on this :warning:** 

## What? Why?

* Closes #13981

This PR introduces support for displaying multiple product images through a reusable carousel component.

As part of the broader multiple-images initiative, this change implements the carousel experience in:
* Shop product details modal (shopper view)
* Admin product preview

The carousel is only displayed when a product has multiple images. Products with a single image continue to show the existing single-image experience.

To support this functionality, this PR:
* Adds support for accessing multiple images on products.
* Updates the admin product images page to list and manage all product images.
* Introduces reusable carousel and thumbnail carousel components with:
  * Previous/next navigation
  * Thumbnail navigation
  * Accessibility support
  * Optional image captions
* Adds carousel support to the admin product preview.
* Adds carousel support to the frontshop product modal.
* Exposes product images through the product API serializer.
* Provides fallback behaviour for products without images.
* Adds component, helper, serializer and frontend test coverage.

This work allows us to build and validate the carousel experience ahead of the future backoffice work for managing multiple images.

### Single image products

* Open a product with a single image in the shop.
* Open the product details modal.
* Verify only the product image is shown.
* Verify no carousel navigation or thumbnails are displayed.

### Multiple image products (shop)

* Open a product with multiple images.
* Open the product details modal.
* Verify a carousel is displayed.
* Verify all product images are available.
* Verify previous/next navigation works.
* Verify thumbnail navigation works.
* Verify images are displayed in upload order (oldest to newest).

### Multiple image products (admin)

* Open **Admin → Product Preview** for a product with multiple images.
* Verify a carousel is displayed.
* Verify all product images are available.
* Verify previous/next navigation works.
* Verify thumbnail navigation works.
* Verify images are displayed in upload order (oldest to newest).

### Image management

* Open **Admin → Product Images**.
* Verify all product images are listed.
* Verify image edit functionality still works.
* Verify image deletion still works.

### Regression testing

* Verify products without images continue to display the placeholder image.
* Verify existing single-image product workflows continue to function as before.

## Release notes

Changelog Category (reviewers may add a label for the release notes):

* [ ] User facing changes
* [ ] API changes (V0, V1, DFC or Webhook)
* [x] Technical changes only
* [ ] Feature toggled

## Dependencies

Need the following PR to be merged along with this one: 
* #14364 

## Documentation updates

None.

### CODE REVIEW

**Comment by chahmedejaz** [comment]
Hey @mariocarabotta - The carousel for multiple images on the product has been deployed to FR Staging. Please review.
I created the data for this product:
https://staging.coopcircuits.fr/admin/products/71817/images

But of course, you may create your own and play around. Thanks.

P.S. Should we introduce a feature flag for this feature, or deploy it all at once?

**Comment by mariocarabotta** [comment]
nice work!! some visual changes below (sorry pretty rough notes, tweaked in the browser).
I need to think about how the images behave when they don't fit the container. I'll have a look at the swiper options, things like resize to fill or similar. also I might make some further visual tweaks later, but hopefully I can push them directly. i've got a local environment set up now.

I will come back to you re feature flag, just checking with Rachel.

**active image container (.ofn-thumbnail-carousel__image)**
background-color: #fff;
border-radius: 8px;

**arrow circle** 
background: rgb(255 255 255);
height: 1.5rem;
width: 1.5rem;

**arrows**
color: #444;
font-weight: 400;
**left arrow only** padding: 0 2px 2px 1px;
**right arrow only** padding: 0 0px 2px 1px;

**caption**
background: rgb(255 255 255);
REMOVE right: 0;
margin-left: 16px;
border-radius: 8px 8px 0px 0px;
font-family: Roboto, Arial, sans-serif;
font-weight: 600;
font-size: 0.9rem;
padding: 8px 12px 4px 12px;
color: #444;

**ofn-thumbnail-carousel__thumbs ng-scope**
REMOVE justify-content: center;

**Comment by RachL** [comment]
@mariocarabotta @chahmedejaz I don't see an advantage to use a feature flag here. However then let's move this PR through testing with testers as well, so we check if there aren't any edge case.

We probably need to update the acceptance criteria:

- there is an item to confirm: _And the order is based on the upload time (TBC)_
- we need to test the scenario of a product having multiple images before the feature is added and the case where new features are added after. Or does it not change anything?

Is there a max. of pictures we can handle?

**Comment by mariocarabotta** [comment]
> @mariocarabotta @chahmedejaz I don't see an advantage to use a feature flag here. However then let's move this PR through testing with testers as well, so we check if there aren't any edge case.
> 
this issue will need to be released together with[ this one](https://github.com/openfoodfoundation/openfoodnetwork/issues/14157), so that people don't see any carousel in production, but we are still able to see and test it in staging. we'll also need to give a heads up to the testers so that they have some test accounts in staging with multiple images. once we hide the button to add, they won't be able to include additional images to a product.

> We probably need to update the acceptance criteria:
> 
> * there is an item to confirm: _And the order is based on the upload time (TBC)_

updated to reflect how it currently behaves (from the oldest upload to the most recent). I think it's ok like this but please let me know if you think otherwise.

> * we need to test the scenario of a product having multiple images before the feature is added and the case where new features are added after. Or does it not change anything?
> 

mmm it's a bit tricky because what we are going to be testing now is a product with multiple product images, but what we are going to enable for users in production eventually is a product with one product image and one image per variant. In my mind this issue is really building the carousel frontend and testing that it can show multiple images, as a UI element. 
[Maybe the issue where we enable to upload variant images](https://github.com/openfoodfoundation/openfoodnetwork/issues/13978) might be the best place to test the scenario you describe?

> Is there a max. of pictures we can handle?

I have been back and forth on specifying this. The max of images would be the same as the max of variants that a product can have. is there a limit to how many variants we can have per product?

In terms of how it renders a considerable number of images in the carousel, we'll need to fix it. [Here's an example](https://github.com/openfoodfoundation/openfoodnetwork/pull/14337). I would prefer if images would just wrap, but I am also aware that some producers have 20 variants per product. But maybe they would not upload a photo for each one. I am leaning towards wrapping rather than a side scrolling which is not particularly accessible. Thoughts?

**Comment by RachL** [comment]
thanks Mario.

> this issue will need to be released together withhttps://github.com/openfoodfoundation/openfoodnetwork/issues/14157

oh I didn't know about this dependency. It sounds to me more complex then than introducing a feature toggle now and releasing carousel only when the multiple image upload is available. Let's see what @chahmedejaz thinks.

If we introduce a feature toggle now; we can leave the carousel hidden and we avoid the work on  https://github.com/openfoodfoundation/openfoodnetwork/issues/14157

> In my mind this issue is really building the carousel frontend and testing that it can show multiple images, as a UI element.

in that case, and again if we have a feature toggle, the easiest here is to merge without testing manually (the feature is hidden, therefore no problem). And wait for a complete test when the image upload is ready.

Aligned on wrapping (but i don't think your example link is working?)

**Comment by chahmedejaz** [comment]
Thanks for the feedback, @mariocarabotta - I’ll address this and follow up.

@RachL, @mariocarabotta For this case, I think we have two options:

1. Introduce a feature flag for this behavior, or
2. Include the work from the other related issue in this PR so we can release the full carousel-related functionality, including prevention, together.

I’m leaning toward the second option since the additional work for the related issue seems relatively small, and it would allow us to ship the carousel changes in a more complete state.

**Comment by mariocarabotta** [comment]
given @chahmedejaz comment, how about we go for his suggested option 2?
I will prepare some more test products in staging australia and will continue to test the UI next week.
We can then let testers do proper end to end testing when [this one](https://github.com/openfoodfoundation/openfoodnetwork/issues/13978) is ready for them.

**Comment by RachL** [comment]
Let's go for option 2 then! Let's just keep in mind that sometimes the follow-up PR takes more time, and maintaining the previous one can be a pain. 
I don't think this will be the case here, so happy to not use the feature toggle. Just mentioning this scenario for future cases: feature toggle allow us to divide the work in several PR. easy to review and merged incrementally.

**Comment by chahmedejaz** [comment]
Adding a note here as well:
- As discussed in the delivery circle yesterday, your design feedback is addressed @mariocarabotta, and it's ready for review.

**Comment by mariocarabotta** [comment]
Thank you @chahmedejaz ! I have been testing on Brave and Chrome desktop, Vanadium (Chrome) on mobile. Not extensive coverage, but enough for now. Once we introduce the ability to upload new images for variants, then testers will re-test the carousel.

1. add margin-bottom: -1px; to caption container to avoid rendering glitch line
Screenshot 2026-06-05 at 9 28 34


2. add margin-bottom: 1rem; to thumbs container to avoid images being too close to overlay border
Screenshot 2026-06-05 at 9 29 52


3. arrows position somehow is overwritten, so it's not applied. please note that values are different between the 2 arrows, sorry for the annoying tweaking of this. I can see you're actually using font icons from our icon font set. With Gaetan and Maikel we are testing using inline svg icons from Material Design. If you want to try that as well, feel free to use [chevron backward and forward](https://fonts.google.com/icons?icon.size=24&icon.color=%23e3e3e3&icon.query=chevron). I'm very sorry I assumed there were some controls as part of the carousel itself, so I didn't mention the new icons. If you prefer to stick with the current implementation, totally fair! These are the values to tweak the current arrows. 
left arrow only padding: 0 2px 2px 1px;
right arrow only padding: 0 0px 2px 1px;
Screenshot 2026-06-05 at 9 26 54


4. For when there are many images: please wrap them instead of scrolling. Adding flex-wrap: wrap; to the thumbs container should do the trick
Screenshot 2026-06-05 at 9 26 14


5. there's a weird line showing up at times. I believe this is a rendering issue where the image at the top is actually not fully covering the one below. It looks like adding padding: 1px; to the .swiper-slide container would fix it
Screenshot 2026-06-05 at 9 22 13


6. active thumb should be 100% opacity, while the others 50%. see below expected result
Screenshot 2026-06-05 at 9 50 53


7. is it possible to enable swiping interactions on mobile, and keyboard controls? it would make it a bit more accessible I believe. There seems to be also a lazy load option that might be helpful (if it's a simple flag implementation)


There's a lot in here, we can also move some of these into a separate issue for later if you prefer. Thank you!

**Comment by chahmedejaz** [comment]
Hey @mariocarabotta - Thanks for the feedback! I tried to address all of it and following is the brief demo.

https://github.com/user-attachments/assets/a5a732d8-bb55-4b06-81e2-c0cc4a26d3a2

Please review the latest changes on the staging and let me know if we require any changes to it. Thanks!

**Comment by mariocarabotta** [comment]
Looking good, thank you for taking the time to address all of it!

The last tweaks (promised!)

- change icons height and width to 24px (thank you so much for replacing them!)
- set border: none; to arrow circles ( the ones with class ofn-thumbnail-carousel__button ofn-thumbnail-carousel__button--prev and ofn-thumbnail-carousel__button ofn-thumbnail-carousel__button--next)
- the circles now are sliding with the images. Is it possible to keep them in place?

thank you so much, this is looking great 🙏🏻

**Comment by chahmedejaz** [comment]
Hey @mariocarabotta, No worries at all, we keep getting better now! 😄

The feedback is addressed, please review and let me know how does it look? Thanks!

https://github.com/user-attachments/assets/a10c1024-c692-4658-a609-0409d5a1b070

**Comment by mariocarabotta** [comment]
> Hey @mariocarabotta, No worries at all, we keep getting better now! 😄
> 
> The feedback is addressed, please review and let me know how does it look? Thanks!
> 
>  20260609-0016-34.6990623.mp4

tested again and I would say now the UI is perfect, amazing work!

I have found a usability issue when on mobile (both testing with the browser dev tools and on 2 actual mobile devices (various versions of chrome). 2 sides of it:

1. when opening on mobile breakpoint, the overlay is scrolled to the image/carousel div. To be honest I'm not sure if this was already there. Also, It might be addressed by Gaetan's work on grid view, since he's rebuilding the overlay. Worth checking with him as well maybe?
2. The carousel div itself has some lock scrolling behaviour, so you have to tap on the edge of the carousel (probably ouside of the div) to scroll up or down and read the rest of the content or close the overlay. I'm surprised I didn't notice this before, I wonder if it comes from the swiping setting in swiper? Or maybe I just didn't have products with long content that needed scrolling..

These 2 combined make it quite hard to use the overlay on mobile. You can try to chek this behavoiur with the topinambur product in the shop linked below. there's also an egg product that shows the first problem without the carousel. I can take a video if you can't replicate it, just let me know.

https://staging.openfoodnetwork.org.au/ofn-shop/shop

**Comment by RachL** [comment]
@mariocarabotta FYI we have a browserstack account (credential in vaultwarden): https://www.browserstack.com/users/sign_in so you don't need to test on actual devices :muscle:

**Comment by chahmedejaz** [comment]
Thanks for catching this, @mariocarabotta! I missed it as well since I didn't have any products with long descriptions. 😅
The issue should now be fixed and has been deployed to AU staging. It's ready for another review.
If you still notice the problem, could you share a short video or screen recording? That would help me investigate further.
Here's the screen recording from my mobile testing:

https://github.com/user-attachments/assets/ebc2334c-4431-4d36-ae2a-22c1a65af807

**Comment by mariocarabotta** [comment]
looking great @chahmedejaz, now it's fixed! 
I don't need to do any further testing, it's looking very good from my side. I would still let testers do some further review though.

amazing work and thank you for your patience

File: app/components/carousel_component/carousel_component.scss [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  left: 0;
+  right: 0;
+  bottom: 0;
+}
+
+%ofn-carousel-button {
+  position: absolute;
+  top: 50%;
+  transform: translateY(-50%);
+  z-index: 2;
```
Comment (rioug) [inline app/components/carousel_component/carousel_component.scss]: I think it's would be good to use a variables here, and defined it in a central place with other `z-index` value like what we started doing for the admin CSS :https://github.com/openfoodfoundation/openfoodnetwork/blob/11e3b90d71eb8f5961e6230c1b7a4dd5c096a74f/app/webpacker/css/admin/globals/variables.scss#L185-L192

File: app/webpacker/css/admin/pages/product_preview.scss
Comment (rioug) [inline app/webpacker/css/admin/pages/product_preview.scss:1]: It looks like you copied the CSS from the frontend, ideally we would share the same CSS so any changes to the frontend flows to the product preview. Could you have a look at sharing the CSS ?

File: app/webpacker/css/darkswarm/all.scss [hunk 2]
```diff
@@ -78,3 +78,5 @@ ofn-modal {
 @import '../admin/shared/scroll_bar';
 
 @import 'modal_component/modal_component';
+@import 'carousel_component/carousel_component';
+@import 'thumbnail_carousel_component/thumbnail_carousel_component';
```
Comment (rioug) [inline app/webpacker/css/darkswarm/all.scss:83]: I don't see any CSS for `thumbnail_carousel_component'` is this needed ?

File: spec/components/carousel_component_spec.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+RSpec.describe CarouselComponent, type: :component do
```
Comment (rioug) [inline spec/components/carousel_component_spec.rb:3]: 💯 !

File: app/assets/javascripts/darkswarm/controllers/products/product_node_controller.js.coffee [hunk 4]
```diff
@@ -2,5 +2,56 @@ angular.module('Darkswarm').controller "ProductNodeCtrl", ($scope, $modal, Filte
   $scope.enterprise = $scope.product.supplier # For the modal, so it's consistent
   $scope.productPropertySelectors = FilterSelectorsService.createSelectors()
 
+  $scope.currentProductImageIndex = 0
+  $scope._productCarouselImages = null
```
Comment (rioug) [inline app/assets/javascripts/darkswarm/controllers/products/product_node_controller.js.coffee]: Why the underscore in the name here ?

File: app/components/carousel_component/carousel_component.scss [hunk 5]
```diff
@@ -11,34 +11,46 @@
...
   left: 0;
-  right: 0;
   bottom: 0;
+  margin-left: 16px;
+  border-radius: 8px 8px 0px 0px;
+  font-family: Roboto, Arial, sans-serif;
+  font-weight: 600;
+  font-size: 0.9rem;
+  padding: 8px 12px 4px 12px;
+  color: #444;
```
Comment (rioug) [inline app/components/carousel_component/carousel_component.scss]: I think in general it would be good to use SCSS variables when applicable,could you could review the final CSS and add variables ?

File: app/assets/javascripts/templates/product_modal.html.haml [hunk 6]
```diff
@@ -15,7 +15,40 @@
...
+      .ofn-thumbnail-carousel__thumbs{"ng-if" => "productCarouselImages().length > 1"}
+        %button.ofn-thumbnail-carousel__thumb{
+          type: "button",
+          "ng-repeat" => "image in productCarouselImages() track by $index",
+          "ng-click" => "selectProductImage($index)",
+          "ng-class" => "{'is-active': $index === currentProductImageIndex}",
+          "ng-attr-aria-label" => "Go to slide {{$index + 1}}",
+          "ng-attr-aria-pressed" => "{{$index === currentProductImageIndex}}"
+        }
+          %img.ofn-thumbnail-carousel__thumb-image{"ng-src" => "{{image.thumb_url || image.url}}", "ng-attr-alt" => "{{image.alt}}"}
```
Comment (rioug) [inline app/assets/javascripts/templates/product_modal.html.haml]: Looking at this and some of the other later commit, it looks like you had to re implement the Carousel functionality in AngularJS. I'd prefer if we could void that, because now we would have two places to look at any time we need to make changes to the carousel. And from what I can see you had to re implement some of the swipping functionality we get out of the box the Stimulus carousel. 
I would much rather we rewrite the modal using StimulusJS and use the same carousel everywhere. I am not sure that's easily feasible, but I think it's worth investigating. 
This PR sort of introduce the StimulusJS product modal ( I just reused the one we use for the product preview in the backoffice. ). The modal is included the html, but we could probably do something similar to what is done in the backoffice, load the modal via an AJAX call. 
@openfoodfoundation/reviewers what do you think ?

Comment (mkllnk) [reply app/assets/javascripts/templates/product_modal.html.haml]: > The modal is included the html, but we could probably do something similar to what is done in the backoffice, load the modal via an AJAX call.

That sounds good. We probably don't want to load all the product modals into the shop list. That would be bad for performance.

File: app/assets/javascripts/darkswarm/controllers/products/product_node_controller.js.coffee (hunk 4 above)
Comment (chahmedejaz) [reply app/assets/javascripts/darkswarm/controllers/products/product_node_controller.js.coffee]: The leading underscore is intentional. It's being used to indicate that this is an internal/private variable rather than a public one. The naming is purely for annotation and readability purposes.

File: app/webpacker/css/darkswarm/all.scss (hunk 2 above)
Comment (chahmedejaz) [reply app/webpacker/css/darkswarm/all.scss:83]: The CSS file exists at `app/components/thumbnail_carousel_component/thumbnail_carousel_component.scss` and is imported in both `darkswarm/all.scss` and `admin/all.scss`.

**Comment by chahmedejaz** [comment]
Hi @roug - Thanks for the feedback. I've addressed it in the following commits. It's ready for review. Thanks

- https://github.com/openfoodfoundation/openfoodnetwork/pull/14337/commits/2cfbecd8ed9352cbffa05fb8430d9cd8e4396ca4
- https://github.com/openfoodfoundation/openfoodnetwork/pull/14337/commits/f6dccf841c04e4edb7171f0ab453c75ee0b495f1

File: app/webpacker/controllers/shop_product_modal_controller.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  async load(url) {
+    if (!url) return;
+
+    const container = document.getElementById("shop-product-modal-container");
+    if (!container) return;
+
+    this.abortPendingRequest();
+    this.abortController = new AbortController();
```
Comment (rioug) [inline app/webpacker/controllers/shop_product_modal_controller.js:23]: TIL about `AbortController` !

File: app/webpacker/controllers/shop_product_modal_controller.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+    if (!this.hasUrlValue) return;
+
+    await this.load(this.urlValue);
+  }
+
+  async load(url) {
+    if (!url) return;
+
+    const container = document.getElementById("shop-product-modal-container");
```
Comment (rioug) [inline app/webpacker/controllers/shop_product_modal_controller.js:19]: I think ideally this would be target associated with the controller, but it's probably too tricky with our existing HMTL.

File: app/views/shop/_product_modal.html.haml [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+              %span= property[:name]
+
+      - if @product.description.present?
+        .product-description{ "data-controller" => "add-blank-to-link" }
+          %p.text-small
+            - # description is sanitized in Spree::Product#description method
+            = @product.description.html_safe
+
+    .columns.small-12.medium-6.large-6.product-img
+      = render ThumbnailCarouselComponent.new(images: @carousel_images, show_captions: true, class: (@product.images.any? ? nil : "placeholder"))
```
Comment (rioug) [inline app/views/shop/_product_modal.html.haml]: It would have been better to reuse the existing template from the backoffice, like what I did here : https://github.com/openfoodfoundation/openfoodnetwork/pull/14345/changes/67316ea8bc0ab379fa00c7fb2c741aa0e3622f13 
That said, it's going to create conflict if we both do the same change : )  Let's leave if as it is, and I'll sort it out while I work on the Product Grid view.

File: spec/support/request/ui_component_helper.rb [hunk 10]
```diff
@@ -55,7 +55,11 @@ def have_login_modal
   end
 
   def open_product_modal(product)
-    page.find("a", text: product.name).click
+    page.find("#product-#{product.id} [data-controller~='shop-product-modal']",
```
Comment (rioug) [inline spec/support/request/ui_component_helper.rb:58]: This would be better if it wasn't tied to the `data-controller` name, can't you use the product link like before ?

**Review by rioug — APPROVED** [review]
Thanks @chahmedejaz , it's much better like this. Maybe we can slightly improve one of spec, but it's otherwise good to go !

**Review by mkllnk — APPROVED** [review]
I have to admit that I looked at this in a rush. Generally it seems okay and Mario verified that it worked. It's a lot more code than I hoped though. I don't have any suggestions for improvements. I guess, I was just hoping that using an existing component would be much easier.

**Comment by AMEA-LYON** [comment]
All good ! Great improvement, congratulations ! 🎉 

Just a remark : sorting images manually is still enabled (see screenshot) and is without effect, I seem to recall that a previous PR was about disabling this feature... Maybe just a merge missing ?

14337_images

**Comment by chahmedejaz** [comment]
Thanks @AMEA-LYON - For the:
> Just a remark : sorting images manually is still enabled (see screenshot) and is without effect, I seem to recall that a previous PR was about disabling this feature... Maybe just a merge missing ?

I think let's skip this for now as we won't be displaying multiple images to the users just now. So, we can fix this in the future PRs. Thanks for raising this though! 👍
